src/pubsub.ts

```diff
@@ -57,56 +59,65 @@ export class AMQPPubSub implements PubSubEngine {
     if (refs && refs.length > 0) {
       const newRefs = [...refs, id];
       this.subsRefsMap[routingKey] = newRefs;
-      return Promise.resolve(id);
-    } else {
-      return this.subscriber.subscribe(this.exchange, routingKey, this.onMessage.bind(this))
-      .then(disposer => {
-        this.subsRefsMap[routingKey] = [
-          ...(this.subsRefsMap[routingKey] || []),
-          id,
-        ];
-        if (this.unsubscribeMap[routingKey]) {
-          return disposer();
-        }
-        this.unsubscribeMap[routingKey] = disposer;
-        return Promise.resolve(id);
-      });
+      return id;
     }
+
+    this.subsRefsMap[routingKey] = [
+      ...(this.subsRefsMap[routingKey] || []),
+      id,
+    ];
```

Comment:
This is the core of the race condition fix. This ref setup should always run before waiting on subscribe so calls that subscribe calls that happen very close after will not overwrite this data or create redundant consumers.